File: crates/rewinds/src/link.rs
//! Serial-link carrier for the desktop host — one pluggable transport that
//! relays the emulator's *opaque* serial frames between two `rewinds` instances.
//!
//! The emulator core is carrier-blind: it emits and consumes frame bytes and
//! knows nothing about how they travel. This TCP transport is one carrier; a
//! Swift Bonjour or web WebRTC host would implement the same length-prefixed
//! relay in its own language against the same frames. Frames are length-prefixed
//! (1-byte length + payload) so the carrier never needs to know their contents.

use std::io::{ErrorKind, Read, Write};
use std::net::{TcpListener, TcpStream};
use std::time::{Duration, Instant};

/// A pluggable serial-link carrier. `send` queues an opaque frame; `recv` flushes
/// pending writes and returns any whole frames that have arrived.
pub trait Transport {
    fn send(&mut self, frame: &[u8]);
    fn recv(&mut self) -> Vec<Vec<u8>>;

    /// Block (up to `timeout`) until at least one frame has arrived, returning all that
    /// are buffered. Empty on timeout. This is the transfer barrier's blocking wait: the
    /// master suspends here until the peer's word for the in-flight round arrives.
    fn recv_blocking(&mut self, timeout: Duration) -> Vec<Vec<u8>> {
        let deadline = Instant::now() + timeout;
        loop {
            let frames = self.recv();
            if !frames.is_empty() {
                return frames;
            }
            if Instant::now() >= deadline {
                return Vec::new();
            }
            std::thread::sleep(Duration::from_micros(100));
        }
    }
}

/// A two-unit TCP link: the parent listens, the child connects. (Three- and
/// four-unit links would extend this into a parent-hub star relay.)
pub struct TcpLink {
    stream: TcpStream,
    out: Vec<u8>,
    inbuf: Vec<u8>,
    id: u8,
    count: u8,
}

impl TcpLink {
    /// Parent: wait for one child to connect on `port`. This unit becomes id 0.
    pub fn listen(port: u16) -> std::io::Result<Self> {
        let listener = TcpListener::bind(("0.0.0.0", port))?;
        log::info!("serial link: waiting for a peer on port {port}…");
        let (stream, addr) = listener.accept()?;
        log::info!("serial link: peer connected from {addr}");
        Self::wrap(stream, 0)
    }

    /// Child: connect to a parent at `addr` (host:port). This unit becomes id 1.
    pub fn connect(addr: &str) -> std::io::Result<Self> {
        let stream = TcpStream::connect(addr)?;
        log::info!("serial link: connected to {addr}");
        Self::wrap(stream, 1)
    }

    fn wrap(stream: TcpStream, id: u8) -> std::io::Result<Self> {
        stream.set_nonblocking(true)?;
        stream.set_nodelay(true).ok();
        Ok(Self { stream, out: Vec::new(), inbuf: Vec::new(), id, count: 2 })
    }

    pub fn id(&self) -> u8 {
        self.id
    }
    pub fn count(&self) -> u8 {
        self.count
    }
}

impl TcpLink {
    /// Flush as much of the outbound queue as the socket will accept (non-blocking).
    fn flush_out(&mut self) {
        while !self.out.is_empty() {
            match self.stream.write(&self.out) {
                Ok(0) => break,
                Ok(n) => {
                    self.out.drain(..n);
                }
                Err(e) if e.kind() == ErrorKind::WouldBlock => break,
                Err(_) => break,
            }
        }
    }

    /// Read whatever is immediately available into the inbound buffer (non-blocking).
    fn drain_available(&mut self) {
        let mut tmp = [0u8; 512];
        loop {
            match self.stream.read(&mut tmp) {
                Ok(0) => break, // peer closed
                Ok(n) => self.inbuf.extend_from_slice(&tmp[..n]),
                Err(e) if e.kind() == ErrorKind::WouldBlock => break,
                Err(_) => break,
            }
        }
    }

    /// Split every complete length-prefixed frame out of the inbound buffer.
    fn split_frames(&mut self) -> Vec<Vec<u8>> {
        let mut frames = Vec::new();
        while let Some(&len) = self.inbuf.first() {
            let len = len as usize;
            if self.inbuf.len() < 1 + len {
                break;
            }
            frames.push(self.inbuf[1..1 + len].to_vec());
            self.inbuf.drain(..1 + len);
        }
        frames
    }
}

impl Transport for TcpLink {
    fn send(&mut self, frame: &[u8]) {
        self.out.push(frame.len() as u8);
        self.out.extend_from_slice(frame);
    }

    fn recv(&mut self) -> Vec<Vec<u8>> {
        self.flush_out();
        self.drain_available();
        self.split_frames()
    }

    /// Block up to `timeout` for at least one frame, **parking** the thread on the socket
    /// (a blocking read with a timeout) rather than busy-polling — so the master waiting
    /// for the peer's word leaves the CPU idle instead of spinning it.
    fn recv_blocking(&mut self, timeout: Duration) -> Vec<Vec<u8>> {
        self.flush_out();
        let ready = self.split_frames();
        if !ready.is_empty() {
            return ready;
        }
        let deadline = Instant::now() + timeout;
        self.stream.set_nonblocking(false).ok();
        let mut tmp = [0u8; 512];
        let frames = loop {
            let remaining = deadline.saturating_duration_since(Instant::now());
            if remaining.is_zero() {
                break Vec::new();
            }
            self.stream.set_read_timeout(Some(remaining)).ok();
            match self.stream.read(&mut tmp) {
                Ok(0) => break Vec::new(), // peer closed
                Ok(n) => {
                    self.inbuf.extend_from_slice(&tmp[..n]);
                    let frames = self.split_frames();
                    if !frames.is_empty() {
                        break frames;
                    }
                }
                Err(_) => break Vec::new(), // timeout or error
            }
        };
        // Restore the non-blocking mode `recv` relies on.
        self.stream.set_nonblocking(true).ok();
        let _ = self.stream.set_read_timeout(None);
        frames
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Frames survive a real loopback round-trip, length-framing intact, in both
    /// directions. Uses an OS-assigned port so the test never collides.
    #[test]
    fn relays_frames_both_ways() {
        let listener = TcpListener::bind(("127.0.0.1", 0)).unwrap();
        let addr = listener.local_addr().unwrap();
        let child_stream = TcpStream::connect(addr).unwrap();
        let (parent_stream, _) = listener.accept().unwrap();
        let mut parent = TcpLink::wrap(parent_stream, 0).unwrap();
        let mut child = TcpLink::wrap(child_stream, 1).unwrap();
        assert_eq!((parent.id(), child.id(), parent.count()), (0, 1, 2));

        let pump = |from: &mut TcpLink, to: &mut TcpLink| -> Vec<Vec<u8>> {
            for _ in 0..100 {
                from.recv(); // flush `from`'s outbound queue onto the wire
                let got = to.recv();
                if !got.is_empty() {
                    return got;
                }
                std::thread::sleep(std::time::Duration::from_millis(1));
            }
            Vec::new()
        };

        parent.send(&[1, 2, 3, 4, 5, 6, 7, 8]);
        assert_eq!(pump(&mut parent, &mut child), vec![vec![1, 2, 3, 4, 5, 6, 7, 8]]);
        child.send(&[9, 10, 11, 12, 13, 14, 15, 16]);
        assert_eq!(pump(&mut child, &mut parent), vec![vec![9, 10, 11, 12, 13, 14, 15, 16]]);
    }

    /// The transfer-barrier primitive: `recv_blocking` returns empty at the timeout when
    /// nothing arrives, and delivers a frame that shows up within the window.
    #[test]
    fn recv_blocking_times_out_then_delivers() {
        let listener = TcpListener::bind(("127.0.0.1", 0)).unwrap();
        let addr = listener.local_addr().unwrap();
        let child_stream = TcpStream::connect(addr).unwrap();
        let (parent_stream, _) = listener.accept().unwrap();
        let mut parent = TcpLink::wrap(parent_stream, 0).unwrap();
        let mut child = TcpLink::wrap(child_stream, 1).unwrap();

        // Nothing sent: it waits out the timeout and returns empty.
        let start = Instant::now();
        assert!(child.recv_blocking(Duration::from_millis(20)).is_empty());
        assert!(start.elapsed() >= Duration::from_millis(20));

        // A frame in flight is delivered well within the timeout.
        parent.send(&[1, 2, 3, 4, 5, 6, 7, 8]);
        parent.recv(); // flush parent's outbound onto the wire
        assert_eq!(
            child.recv_blocking(Duration::from_millis(500)),
            vec![vec![1, 2, 3, 4, 5, 6, 7, 8]]
        );
    }
}

File: rewinds-ios/RewindsKit/Sources/RewindsKit/Link.swift
import Foundation
import Network
import Combine
#if canImport(UIKit)
import UIKit
#endif

/// A pluggable serial-link carrier. Mirrors the Rust reference host's `Transport` trait
/// (`crates/rewinds/src/link.rs`): `send` queues one opaque frame; `poll` returns whole
/// frames that have arrived. The emulator core is carrier-blind — it only produces and
/// consumes these opaque `LinkFrame` bytes, and neither end knows how they travel.
protocol LinkCarrier: AnyObject {
    func send(_ frame: Data)
    func poll() -> [Data]
}

/// Bonjour service identity for RewinDS serial links, shared by the advertiser (host) and
/// the browser (joiner).
enum LinkService {
    static let type = "_rewinds-link._tcp"
    // A link connects two *compatible systems* and lets the games decide if they recognize
    // each other — exactly like a real cable. So we gate on console family + wire/ABI
    // version, never on ROM identity (Ruby↔Sapphire, FireRed↔LeafGreen, and cross-language
    // trades all use different ROMs and must link). `rom` is display-only.
    static let txtConsole = "con" // "gba" / "nds" — must match to link
    static let txtVersion = "ver" // the core ABI version (rewinds_core_version) — must match
    static let txtRom = "rom"     // the ROM's display name — informational only

    /// The name this device advertises itself under. On recent iOS this is a generic
    /// "iPhone" without a special entitlement, which is fine — the ROM in the TXT record is
    /// the useful disambiguator.
    static var deviceName: String {
        #if canImport(UIKit)
        return UIDevice.current.name
        #else
        return Host.current().localizedName ?? "RewinDS"
        #endif
    }

    static func value(_ txt: NWTXTRecord, _ key: String) -> String? {
        if case let .string(v) = txt.getEntry(for: key) { return v }
        return nil
    }
}

/// A framed serial-link connection over one `NWConnection`. The wire format matches the
/// Rust reference host's `TcpLink`: each frame is length-prefixed with a single byte
/// (`[len][payload…]`). Every carrier path uses this — the dev TCP connect and both sides
/// of a Bonjour link — differing only in how the connection is created and which unit id
/// it takes (0 = parent/host, 1 = child/joiner).
///
/// # Threading
/// The `NWConnection` runs on its own queue; its callbacks touch only this object's
/// `lock`-guarded state, never the emulator core. `send`/`poll`/`currentStatus` are safe
/// to call from the emulation thread.
final class LinkConnection: LinkCarrier {
    enum Status: Equatable {
        case connecting
        case ready
        case failed(String)
        case closed
    }

    let id: UInt8
    let count: UInt8 = 2

    private let connection: NWConnection
    private let queue = DispatchQueue(label: "com.setimmediate.link")
    // A condition (not a plain lock) so `recvBlocking` can park the emulation thread on
    // the socket at a transfer barrier and be woken the instant a frame or a status change
    // arrives — no busy-polling.
    private let cond = NSCondition()
    private var status: Status = .connecting
    private var inbox: [Data] = []
    private var inbuf = [UInt8]()

    /// Connect out to a host:port (the dev TCP path). This unit is the child.
    convenience init(host: String, port: UInt16, id: UInt8) {
        self.init(connection: NWConnection(
            host: NWEndpoint.Host(host),
            port: NWEndpoint.Port(rawValue: port) ?? .any,
            using: Self.params()), id: id)
    }

    /// Connect out to a discovered Bonjour endpoint. This unit is the child (id 1).
    convenience init(endpoint: NWEndpoint, id: UInt8) {
        self.init(connection: NWConnection(to: endpoint, using: Self.params()), id: id)
    }

    /// Adopt a connection — an inbound one accepted by a listener (host, id 0), or one of
    /// the outbound connections above — starting it and pumping its receive loop.
    init(connection: NWConnection, id: UInt8) {
        self.connection = connection
        self.id = id
        connection.stateUpdateHandler = { [weak self] state in self?.onState(state) }
        connection.start(queue: queue)
        receiveNext()
    }

    private static func params() -> NWParameters {
        let tcp = NWProtocolTCP.Options()
        tcp.noDelay = true
        let params = NWParameters(tls: nil, tcp: tcp)
        params.includePeerToPeer = true
        return params
    }

    /// Tear down the connection. Safe to call from any thread.
    func cancel() {
        connection.cancel()
        setStatus(.closed)
    }

    func currentStatus() -> Status {
        cond.lock(); defer { cond.unlock() }
        return status
    }

    // --- LinkCarrier ----------------------------------------------------------

    func send(_ frame: Data) {
        guard frame.count <= 255 else { return } // one-byte length prefix
        var packet = Data([UInt8(frame.count)])
        packet.append(frame)
        connection.send(content: packet, completion: .contentProcessed { _ in })
    }

    func poll() -> [Data] {
        cond.lock(); defer { cond.unlock() }
        guard !inbox.isEmpty else { return [] }
        let frames = inbox
        inbox.removeAll(keepingCapacity: true)
        return frames
    }

    // --- Connection plumbing (carrier queue) ----------------------------------

    private func onState(_ state: NWConnection.State) {
        switch state {
        case .ready:
            setStatus(.ready)
        case .failed(let err), .waiting(let err):
            // `waiting` means the endpoint is unreachable (e.g. connection refused); for a
            // deliberate connect, surface it as a failure rather than retrying forever.
            setStatus(.failed(err.localizedDescription))
        case .cancelled:
            setStatus(.closed)
        default:
            break
        }
    }

    private func receiveNext() {
        connection.receive(minimumIncompleteLength: 1, maximumLength: 8192) { [weak self] data, _, isComplete, error in
            guard let self else { return }
            if let data, !data.isEmpty { self.ingest(data) }
            if let error { self.setStatus(.failed(error.localizedDescription)); return }
            if isComplete { self.setStatus(.closed); return }
            self.receiveNext()
        }
    }

    private func ingest(_ data: Data) {
        cond.lock()
        inbuf.append(contentsOf: data)
        // Split off every complete length-prefixed frame.
        var gotFrame = false
        while let len = inbuf.first.map(Int.init), inbuf.count >= 1 + len {
            inbox.append(Data(inbuf[1..<(1 + len)]))
            inbuf.removeFirst(1 + len)
            gotFrame = true
        }
        if gotFrame { cond.signal() } // wake a barrier wait in `recvBlocking`
        cond.unlock()
    }

    private func setStatus(_ newStatus: Status) {
        cond.lock()
        status = newStatus
        cond.signal() // wake a barrier wait so a drop/close doesn't wait out the timeout
        cond.unlock()
    }

    /// Block up to `timeout` for at least one frame, **parking** the emulation thread on
    /// the socket rather than spinning — the transfer barrier's wait. Returns whatever
    /// frames arrived (empty on timeout or a dead connection).
    func recvBlocking(timeout: TimeInterval) -> [Data] {
        cond.lock(); defer { cond.unlock() }
        let deadline = Date().addingTimeInterval(timeout)
        while inbox.isEmpty && Self.waitable(status) && Date() < deadline {
            if !cond.wait(until: deadline) { break } // timed out
        }
        let frames = inbox
        inbox.removeAll(keepingCapacity: true)
        return frames
    }

    /// Whether it's still worth waiting on this connection (not failed/closed).
    private static func waitable(_ status: Status) -> Bool {
        switch status {
        case .connecting, .ready: return true
        case .failed, .closed: return false
        }
    }
}

/// Owns the serial-link carrier and shuttles frames between it and the emulator core in
/// lockstep with the run loop. Hosts (advertises + accepts a peer), joins a discovered
/// peer, or connects to a dev TCP host.
///
/// # Threading
/// The active carrier's whole lifecycle lives on the emulation thread: every entry point
/// records an *intent* (a carrier to adopt, or a disconnect) under `intentLock`, and
/// `pump(_:)` — called from `EmulatorSession.tick()` under the core lock — applies it and
/// reconciles the core's link config. So the core is only ever touched from its owning
/// thread. The `NWListener` (advertising) and its accept callback never touch the core;
/// on accept they just record a carrier intent. `state` is published for the UI on the
/// main queue.
public final class LinkController: ObservableObject {
    public enum State: Equatable {
        case offline
        case hosting             // advertising, waiting for a peer
        case connecting
        case linked
        case disconnected(String)
    }

    @Published public private(set) var state: State = .offline

    public init() {}

    private let intentLock = NSLock()
    private var pendingCarrier: LinkConnection?
    private var pendingDisconnect = false

    // Touched only on the emulation thread (inside `pump`), except `shutdown`.
    private var carrier: LinkConnection?
    private var configured = false

    // Advertising side. Owned here; its accept handler records a carrier intent.
    private var listener: NWListener?
    private static let controlQueue = DispatchQueue(label: "com.setimmediate.link.control")

    // --- UI thread ------------------------------------------------------------

    /// Advertise this device over Bonjour and wait for a peer to join. This unit becomes
    /// the parent (id 0) when a peer connects.
    public func host(rom: String, console: String) {
        LinkNotifier.requestAuthIfNeeded()
        stopListener()
        guard let listener = try? NWListener(using: LinkConnection.hostParams()) else {
            publish(.disconnected("Couldn't start hosting"))
            return
        }
        let txt = NWTXTRecord([LinkService.txtConsole: console,
                               LinkService.txtVersion: String(EmulatorCore.coreVersion),
                               LinkService.txtRom: rom])
        listener.service = NWListener.Service(name: LinkService.deviceName, type: LinkService.type, txtRecord: txt)
        listener.newConnectionHandler = { [weak self] connection in
            guard let self else { connection.cancel(); return }
            // Two-unit only: adopt the first peer as parent (id 0) and stop advertising.
            self.setCarrierIntent(LinkConnection(connection: connection, id: 0))
            self.stopListener()
        }
        listener.stateUpdateHandler = { [weak self] state in
            if case .failed = state { self?.publish(.disconnected("Hosting failed")) }
        }
        self.listener = listener
        listener.start(queue: Self.controlQueue)
        publish(.hosting)
    }

    /// Join a peer discovered over Bonjour. This unit becomes the child (id 1).
    public func join(_ peer: DiscoveredPeer) {
        LinkNotifier.requestAuthIfNeeded()
        stopListener()
        setCarrierIntent(LinkConnection(endpoint: peer.endpoint, id: 1))
        publish(.connecting)
    }

    /// Connect to a dev TCP host (`rewinds --link listen <port>`). This unit is the child.
    public func connectDev(host: String, port: UInt16) {
        stopListener()
        setCarrierIntent(LinkConnection(host: host, port: port, id: 1))
        publish(.connecting)
    }

    /// Request teardown of the current link (and stop advertising).
    public func disconnect() {
        stopListener()
        intentLock.lock()
        pendingDisconnect = true
        pendingCarrier = nil
        intentLock.unlock()
    }

    /// Tear the carrier down synchronously. Call only once the emulation thread has stopped
    /// (from `EmulatorSession.teardown`), so there is no concurrent `pump`.
    func shutdown() {
        stopListener()
        carrier?.cancel()
        carrier = nil
        configured = false
    }

    private func setCarrierIntent(_ connection: LinkConnection) {
        intentLock.lock()
        pendingCarrier = connection
        pendingDisconnect = false
        intentLock.unlock()
    }

    private func stopListener() {
        listener?.cancel()
        listener = nil
    }

    // --- Emulation thread (under the core lock) -------------------------------

    /// Apply pending intents and reconcile the core's link config. Returns whether a link
    /// is up and this frame should be driven by [`runLinkedFrame`] (the transfer barrier);
    /// `false` means run a normal frame (unlinked, or still connecting).
    func reconcile(_ core: EmulatorCore) -> Bool {
        applyIntents(core)
        guard let carrier else { return false }
        switch carrier.currentStatus() {
        case .connecting:
            return false
        case .ready:
            if !configured {
                core.setLinkConfig(connected: true, id: carrier.id, count: carrier.count)
                configured = true
                publish(.linked)
                LinkNotifier.notify("RewinDS linked", "Connected for multiplayer.")
            }
            return true
        case .failed(let message):
            let wasLinked = configured
            teardown(core)
            publish(.disconnected(message))
            if wasLinked { LinkNotifier.notify("Link disconnected", message) }
            return false
        case .closed:
            let wasLinked = configured
            teardown(core)
            publish(.disconnected("Link closed"))
            if wasLinked { LinkNotifier.notify("Link disconnected", "The other player left.") }
            return false
        }
    }

    /// Drive one video frame under the serial transfer barrier, mirroring the desktop host:
    /// advance in short slices, service the carrier between each, and at every transfer send
    /// our word and **park** for the peer's reply before resuming — so each transfer
    /// completes in-frame (clock-locked) even during the bursts a link session fires within
    /// a single frame. A stalled peer (no reply within the timeout) ends the frame early.
    /// Only called when [`reconcile`] returned `true`; runs on the emulation thread.
    func runLinkedFrame(_ core: EmulatorCore) {
        guard let carrier else { return }
        let sliceCycles: UInt64 = 512
        let barrierTimeout: TimeInterval = 0.05
        while true {
            for frame in carrier.poll() { core.linkDeliver(frame) }
            while let frame = core.linkPollOut() { carrier.send(frame) }
            switch core.runStep(maxCycles: sliceCycles) {
            case .frameComplete:
                return
            case .yielded:
                continue
            case .linkPending:
                while let frame = core.linkPollOut() { carrier.send(frame) }
                let frames = carrier.recvBlocking(timeout: barrierTimeout)
                if frames.isEmpty { return } // peer stalled — bail this frame rather than hang
                for frame in frames { core.linkDeliver(frame) }
            }
        }
    }

    private func applyIntents(_ core: EmulatorCore) {
        intentLock.lock()
        let newCarrier = pendingCarrier; pendingCarrier = nil
        let disconnect = pendingDisconnect; pendingDisconnect = false
        intentLock.unlock()

        if disconnect {
            teardown(core)
            publish(.offline)
        }
        if let newCarrier {
            teardown(core)
            carrier = newCarrier
        }
    }

    private func teardown(_ core: EmulatorCore) {
        carrier?.cancel()
        carrier = nil
        if configured {
            core.setLinkConfig(connected: false, id: 0, count: 1)
            configured = false
        }
    }

    private func publish(_ newState: State) {
        DispatchQueue.main.async { [weak self] in self?.state = newState }
    }
}

extension LinkConnection {
    /// Parameters for the advertising listener (TCP, peer-to-peer enabled).
    static func hostParams() -> NWParameters { params() }
}
